### PR TITLE
Change WindowBatchIterator to be a pipeline breaker

### PR DESCRIPTION
**Pull request by mfussenegger**
## Summary of the changes / Why this improves CrateDB

This integrates the ordering logic into the WindowBatchIterator to
simplify the framing logic as it can be done on a List instead of a
source BatchIterator.

The assumption was that this should improve performance as it would
avoid extra row-allocations. (Before we had 1 full materialization in a
extra ordering BatchIterator + additional materializations in the
WindowBatchIterator; Even without ORDER BY the WindowBatchIterator was
materializing all rows due to the peer logic)

But benchmarks indicate that the performance is roughly the same:

    Q: select avg("adRevenue") over (partition by "cCode" order by "visitDate")
                    from uservisits
    C: 1
    | Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
    |   V1    |     9261.637 ± 1215.407 |   8200.260 |   8569.312 |  10257.467 |  12525.779 |
    |   V2    |     9123.659 ±  916.305 |   8147.050 |   8643.932 |   9782.795 |  11572.873 |
    mean:   -   1.50%
    median: +   0.87%

The reason for this is probably that the sort phase takes up the majority of the time. (Profiling seems to confirm this assumption). We could further optimize this by parallelizing the sort phase. (Which is easier on a materialized list than on a lazy source)

## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: sql/src/main/java/io/crate/execution/engine/window/WindowFrameState.java [hunk 1]
```diff
@@ -30,9 +30,11 @@
  */
 public final class WindowFrameState {
 
-    private final int lowerBound;
-    private final int upperBoundExclusive;
-    private List<Object[]> rows;
+    private final List<Object[]> rows;
+    private int lowerBound;
+    private int upperBoundExclusive;
+    private int pStart;
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/engine/window/WindowFrameState.java]: shall we use more descriptive name? "partitionStart/End" should do the trick

File: sql/src/main/java/io/crate/execution/engine/sort/Comparators.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        }
+        return arrayOrdering(positions, orderBy.reverseFlags(), orderBy.nullsFirst());
+    }
+
+    private static <T extends CollectExpression<Row, ?>> Comparator<Object[]> createComparatorWithEval(
+        Supplier<InputFactory.Context<T>> createInputFactoryCtx,
+        OrderBy orderBy) {
+
+        var orderBySymbols = orderBy.orderBySymbols();
+        List<Comparator<Object[]>> comparators = new ArrayList<>(orderBySymbols.size());
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/engine/sort/Comparators.java]: nitpick - can ba ArrayList

File: dex/src/main/java/io/crate/breaker/RowAccounting.java [hunk 3]
```diff
@@ -37,7 +37,7 @@
      * @throws CircuitBreakingException if too much memory would be consumed
      *         after materializing this row.
      */
-    void accountForAndMaybeBreak(Row row);
+    Row accountForAndMaybeBreak(Row row);
```
Comment (andreidan) [inline dex/src/main/java/io/crate/breaker/RowAccounting.java]: This is a bit non-intuititve. Do we absolutely need it? I get that it's useful for pipe-ing the Row to further API calls but I'm not sure that benefit outweighs the possible confustion. An interface like this indicates it'll return a new Row - and given the name of the method, I'd expect to maybe return a total of how much memory it accounted for.

File: sql/src/main/java/io/crate/execution/engine/window/WindowFunctionBatchIterator.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                .thenApply(rows -> computeWindowFunctions(
+                    rows,
+                    cmpPartitionBy,
+                    cmpOrderBy,
+                    numCellsInSourceRow,
+                    windowFunctions,
+                    argsExpressions,
+                    args
+                ))
+                .thenApply(rows -> Lists.transform(rows, Buckets.arrayToSharedRow()::apply)),
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/engine/window/WindowFunctionBatchIterator.java]: ah yes, love this new API composition 👏🏻

File: sql/src/main/java/io/crate/execution/engine/window/WindowFunctionBatchIterator.java [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        if (rows.isEmpty()) {
+            return rows;
+        }
+        maybeSortRows(rows, cmpPartitionBy, cmpOrderBy);
+        int start = 0;
+        int end = rows.size();
+        int pStart = start;
+        int pEnd = findFirstNonPeer(rows, pStart, end, cmpPartitionBy);
+        var frame = new WindowFrameState(pStart, pEnd, rows);
+        boolean isTraceEnabled = LOGGER.isTraceEnabled();
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/engine/window/WindowFunctionBatchIterator.java]: can be static final

Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/window/WindowFunctionBatchIterator.java]: `isTraceEnabled` can change at runtime ( via `SET GLOBAL 'logger...`) . I do the assignment outside the inner loop as a trade-off between having the logging change visible between statement execution vs. *within* a statement being executed.

File: dex/src/main/java/io/crate/breaker/RowAccounting.java (hunk 3 above)
Comment (mfussenegger) [reply dex/src/main/java/io/crate/breaker/RowAccounting.java]: Was a bit skeptical myself. Will change it back.

**Comment by mfussenegger** [comment]
@seut anything to add?

**Pull request merged**